modules/core/src/copy.cpp

```diff
@@ -149,6 +165,130 @@ copyMask_<ushort>(const uchar* _src, size_t sstep, const uchar* mask, size_t mst
     }
 }
 
+#if CV_SIMD_PERMUTE >= 1
+template<> void
+copyMask_<Vec3b>(const uchar* src, size_t sstep, const uchar* mask, size_t mstep, uchar* dst, size_t dstep, Size size)
+{
+    // Hopefully the compiler figures out, and stores the permute as a table.
+    uint8_t CV_DECL_ALIGNED(CV_SIMD_WIDTH) permv[v_uint8::nlanes * 3];
+    for(int x = 0; x < v_uint8::nlanes * 3; x += 3)
+        permv[x] = permv[x + 1] = permv[x + 2] = x / 3;
+
+    for( ; size.height--; mask += mstep, src += sstep, dst += dstep )
+    {
+        int x = 0;
+        int mx = 0;
+        {
+            v_uint8 v_zero = vx_setzero_u8();
+            int stride = v_uint8::nlanes * 3;
+            int len0 = ((size.width * 3) / stride) * stride;
+
+            v_uint8 p0 = vx_load(permv + v_uint8::nlanes * 0);
+            v_uint8 p1 = vx_load(permv + v_uint8::nlanes * 1);
+            v_uint8 p2 = vx_load(permv + v_uint8::nlanes * 2);
+
+            for( ; x < len0; x += stride, mx += v_uint8::nlanes )
+            {
+
+                v_uint8 m8v = vx_load(mask + mx) == v_zero;
+                v_uint8 src0 = vx_load(src + x);
+                v_uint8 src1 = vx_load(src + x + v_uint8::nlanes);
+                v_uint8 src2 = vx_load(src + x + v_uint8::nlanes * 2);
+                v_uint8 dst0 = vx_load(dst + x);
+                v_uint8 dst1 = vx_load(dst + x + v_uint8::nlanes);
+                v_uint8 dst2 = vx_load(dst + x + v_uint8::nlanes * 2);
+
+                // Use a permute to "expand" a value of 0 or -1.
+                v_uint8 m24v0 = v_permute(p0, m8v);
+                v_uint8 m24v1 = v_permute(p1, m8v);
+                v_uint8 m24v2 = v_permute(p2, m8v);
```

Comment:
As I mentioned in #15613 lets introduce more specific intrinsic for that.
- manipulation with control parameter in implementations doesn't looks good (it is not universal and generic) - this way is not optimal for processing of that task
- take a look on `v_zip()` - it interleaves two registers. Here we need interleaving of three registers.
- lets try to avoid conditional compilation
- we already have `v_store_interleave()`. Here we don't need "storing". We need just interleaving from registers into registers. But...
- but even interleaving would not work with masks in the future (see OE-29 - AVX512): there are plans to rework masks support completely (masks type will be a separate "abstract" type with reduced amount of operations).
- so, finally, it is better to introduce here more specific intrinsic: `void v_select3(mask, a1, b1, c1, a2, b2, c2, dst_a, dst_b, dst_c)` (merge these v_permute + v_select operations). It can be implemented with optimal performance for each architecture.

Reply:
I think updating this patch would be highly contingent on the choose of HAL API exposed for masks. I can only assume this would heavily bias towards AVX512. AFAIK composite types (e.g 8u3) don't map onto the HAL.

Maybe `v_memmove_triplet(dstp, srcp, maskp, ntriplets)`?

Reply:
I'm looking at the OpenCV code and I can see uses for a generic v_permute (or v_shuffle) instruction(s) in other areas.  v_permute would need both single and double input registers (besides the control vector).  Without these instructions, work in these areas might not be able to be done.

Reply:
I am still waiting on community feedback before refactoring this patchset.

I think v_permute is well defined for most architectures. I can drop copymask patch while consensus is being developed for masked operations. Other than my suggestion above, I lack insight into what a good HAL would look like for such operations.

Reply:
> v_memmove_triplet

Variant with memory pointers make sense too. For example, skip data loading from source if `v_check_any(mask) == false` or `v_check_all(mask) == true` (blending is not needed).

> ntriplets

Number of interleaved channels?

> composite types (e.g 8u3) don't map onto the HAL

Right, there is no dedicated type for that.
Interleaved data for 3 channels is stored in 3 vector registers. In case of 128-bit int32:
- `{ R1 G1 B1 R2}    {G2 B2 R3 G3}    {B3 R4 G4 B4}`
- see also `v_load_deinterleave()`